src/gpuarray_array_blas.c

```diff
@@ -383,3 +384,191 @@ int GpuArray_rger(double alpha, GpuArray *X, GpuArray *Y, GpuArray *A,
     GpuArray_clear(&copyY);
   return err;
 }
+
+int GpuArray_rgemmBatch_3d(cb_transpose transA, cb_transpose transB, double alpha,
+                           GpuArray *A, GpuArray *B, double beta, GpuArray *C,
+                           int nocopy) {
+  GpuArray *Ap = A;
+  GpuArray copyA;
+  GpuArray *Bp = B;
+  GpuArray copyB;
+  GpuArray *Cp = C;
+  gpuarray_blas_ops *blas;
+  void *ctx;
+  size_t elsize;
+  size_t batchCount, m, n, k, lda, ldb, ldc;
+  cb_order o;
+  int err;
+  gpudata **A_datas = NULL, **B_datas = NULL, **C_datas = NULL;
+  size_t *A_offsets = NULL, *B_offsets = NULL, *C_offsets = NULL;
+
+  if (A->typecode != GA_FLOAT && A->typecode != GA_DOUBLE)
+    return GA_INVALID_ERROR;
+
+  if (A->nd != 3 || B->nd != 3 || C->nd != 3 ||
+      A->typecode != A->typecode || B->typecode != A->typecode ||
+      C->typecode != A->typecode)
+    return GA_VALUE_ERROR;
+
+  if (!(A->flags & GA_ALIGNED) || !(B->flags & GA_ALIGNED) ||
+      !(C->flags & GA_ALIGNED))
+    return GA_UNALIGNED_ERROR;
+
+  batchCount = A->dimensions[0];
+  if (B->dimensions[0] != batchCount || C->dimensions[0] != batchCount)
+    return GA_VALUE_ERROR;
+
+  if (transA == cb_no_trans) {
+    m = A->dimensions[1];
+    k = A->dimensions[2];
+  } else {
+    m = A->dimensions[2];
+    k = A->dimensions[1];
+  }
+
+  if (transB == cb_no_trans) {
+    n = B->dimensions[2];
+    if (B->dimensions[1] != k)
+      return GA_VALUE_ERROR;
+  } else {
+    n = B->dimensions[1];
+    if (B->dimensions[2] != k)
+      return GA_VALUE_ERROR;
+  }
+
+  if (C->dimensions[1] != m || C->dimensions[2] != n)
+    return GA_VALUE_ERROR;
+
+  elsize = gpuarray_get_elsize(A->typecode);
+
+  if (!GpuArray_ISONESEGMENT(A)) {
+    if (nocopy)
+      return GA_COPY_ERROR;
+    else {
+      err = GpuArray_copy(&copyA, A, GA_F_ORDER);
+      if (err != GA_NO_ERROR)
+	goto cleanup;
+      Ap = &copyA;
+    }
+  }
```

Comment:
This check is ok, if a bit restrictive.  It should be ok to have a non-contiguous A as long as the last two dims are.  If you don't want to do this for now that's ok but please add a comment saying that for future reference.